
 # Dealing With Null Values Solution Notebook.
 <br>
 <br>
 It's time to get your hands dirty dude!!
 -Bit mL

# Import Libraries

In [2]:
import numpy as np
import pandas as pd


 Python list <b>headers</b> containing name of headers 


In [3]:
# These are the headers which you need to link to your dataset.
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

# Import Datasets

Use the Pandas method <b>read_csv()</b> to load the data from the web address. Set the parameter  "names" equal to the Python list "headers".


In [4]:
df = pd.read_csv('car_dataset.data', names = headers)

 Use the method <b>head()</b> to display the first five rows of the dataframe. 


In [5]:
# To see what the data set looks like, we'll use the head() method.
df.head(15)


symboling normalized-losses         make fuel-type aspiration  \
0           3                 ?  alfa-romero       gas        std   
1           3                 ?  alfa-romero       gas        std   
2           1                 ?  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2                 ?         audi       gas        std   
6           1               158         audi       gas        std   
7           1                 ?         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0                 ?         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1                 ?          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

As we can see, several question marks appeared in the dataframe; those are missing values which may hinder our further analysis. 

<div>So, how do we identify all those missing values and deal with them?</div> 

<b>How to work with missing data?</b>

Steps for working with missing data:

<ol>
    <li>Identify missing data</li>
    <li>deal with missing data</li>
    <li>correct data format</li>
</ol>


<h1 id="identify_handle_missing_values">Identify and handle missing values</h1>

<h2 id="identify_missing_values">Identify missing values</h2>
<h3>Convert "?" to NaN</h3>
In the car dataset, missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Here we use the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B


In [6]:
# replace "?" to NaN
df.replace ("?", np.nan,inplace=True)



In [7]:
df.head(15)

symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           3               NaN  alfa-romero       gas        std   
2           1               NaN  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2               NaN         audi       gas        std   
6           1               158         audi       gas        std   
7           1               NaN         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0               NaN         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1               NaN          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

Identify_missing_values

<h2>Evaluating for Missing Data</h2>

The missing values are converted to default. We use the following functions to identify these missing values. There are two methods to detect missing data:

<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.


In [8]:
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

"True" stands for missing value, while "False" stands for not missing value.


In [12]:
df.notnull().sum()

symboling            205
normalized-losses    164
make                 205
fuel-type            205
aspiration           205
num-of-doors         203
body-style           205
drive-wheels         205
engine-location      205
wheel-base           205
length               205
width                205
height               205
curb-weight          205
engine-type          205
num-of-cylinders     205
engine-size          205
fuel-system          205
bore                 201
stroke               201
compression-ratio    205
horsepower           203
peak-rpm             203
city-mpg             205
highway-mpg          205
price                201
dtype: int64

In [13]:
missing_data = df.isnull()
missing_data.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

<h2>Count missing values in each column</h2>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values. 
</p>


In [14]:
missing_data = df.isnull()
for column in headers:
    print(column)
    print(missing_data[column].value_counts())
    print("")

symboling
symboling
False    205
Name: count, dtype: int64

normalized-losses
normalized-losses
False    164
True      41
Name: count, dtype: int64

make
make
False    205
Name: count, dtype: int64

fuel-type
fuel-type
False    205
Name: count, dtype: int64

aspiration
aspiration
False    205
Name: count, dtype: int64

num-of-doors
num-of-doors
False    203
True       2
Name: count, dtype: int64

body-style
body-style
False    205
Name: count, dtype: int64

drive-wheels
drive-wheels
False    205
Name: count, dtype: int64

engine-location
engine-location
False    205
Name: count, dtype: int64

wheel-base
wheel-base
False    205
Name: count, dtype: int64

length
length
False    205
Name: count, dtype: int64

width
width
False    205
Name: count, dtype: int64

height
height
False    205
Name: count, dtype: int64

curb-weight
curb-weight
False    205
Name: count, dtype: int64

engine-type
engine-type
False    205
Name: count, dtype: int64

num-of-cylinders
num-of-cylinders
False    205
Nam

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:

<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>


<h2 id="deal_missing_values">Deal with missing data</h2>
<br>
<h3>How to deal with missing data?</h3>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
    </li>
</ol>


Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>

<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>

<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>

<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>


<h3>Calculate the average of the normalized-losses column </h3>


In [15]:
avg_num=df["normalized-losses"].astype("float").mean(axis=0)
print("Average of normalized-losses:", avg_num)

Average of normalized-losses: 122.0


<h3>Replace "NaN" by mean value in "normalized-losses" column</h3>


In [17]:
df["normalized-losses"].replace(np.nan, avg_num, inplace=True)
df.head(15)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_26676\109706684.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["normalized-losses"].replace(np.nan, avg_num, inplace=True)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0             122.0         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

<h3>Calculate the mean value for 'bore' column</h3>


In [24]:
avg_bore=df["bore"].astype("float").mean(axis=0)
print("Average of bore:", avg_bore)


Average of bore: 5.645268292682926


<h3>Replace NaN by mean value for bore column</h3>


In [21]:
df['bore'].replace(np.nan, avg_num, inplace=True)
df.head(15)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_26676\177304721.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['bore'].replace(np.nan, avg_num, inplace=True)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0             122.0         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

<h3>Calculate the mean value for 'stroke' column</h3>

In [ ]:
# Write your code below and press Shift+Enter to execute 
avg_stroke=df["stroke"].astype("float").mean(axis=0)
print("Average of stroke:", avg_stroke)


Average of stroke: 5.572390243902437


<h3>Replace NaN by mean value for stroke column</h3>

In [26]:
df['stroke'].replace(np.nan, avg_num, inplace=True)
df.head(15)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_26676\2105432805.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['stroke'].replace(np.nan, avg_num, inplace=True)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0             122.0         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

<details><summary>Click here for the solution</summary>

```python
#Calculate the mean vaule for "stroke" column
avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Average of stroke:", avg_stroke)

# replace NaN by mean value in "stroke" column
df["stroke"].replace(np.nan, avg_stroke, inplace = True)
```

</details>


<h4>Calculate the mean value for the  'horsepower' column:</h4>


In [28]:
avg_horsepower=df["horsepower"].astype("float").mean(axis=0)
print("Average of horsepower:", avg_horsepower)

Average of horsepower: 103.2941001784652


<h4>Replace "NaN" by mean value:</h4>


In [27]:
df['horsepower'].replace(np.nan, avg_num, inplace=True)
df.head(15)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_26676\607169591.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['horsepower'].replace(np.nan, avg_num, inplace=True)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0             122.0         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

<h4>Calculate the mean value for 'peak-rpm' column:</h4>


In [29]:
avg_peakrpm=df["peak-rpm"].astype("float").mean(axis=0)
print("Average of peak-rpm:", avg_peakrpm)

Average of peak-rpm: 5125.369458128079


<h4>Replace NaN by mean value:</h4>


In [30]:
df['peak-rpm'].replace(np.nan, avg_num, inplace=True)
df.head(15) 

C:\Users\ASUS\AppData\Local\Temp\ipykernel_26676\2215090774.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['peak-rpm'].replace(np.nan, avg_num, inplace=True)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0             122.0         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

<br>

## Alright you have done great job till now, but its not finished yet dude!!!
<br>

<ul>
<li>Now we have done for replacing with mean.</li>
<li>Its time to replace null values with most frequent values </li>
<li>We will do this for num-of-doors column</li>
</ul>

<h4>Find the most frequent value for num of doors</h4>

In [31]:
df['num-of-doors'].value_counts()
df['body-style'].value_counts()

body-style
sedan          96
hatchback      70
wagon          25
hardtop         8
convertible     6
Name: count, dtype: int64

<h4>We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:</h4>


In [32]:
df['num-of-doors'].value_counts().idxmax()

'four'

<h4>The replacement procedure is very similar to what we have seen previously</h4>


In [33]:
#replace the missing 'num-of-doors' values by the most frequent 
df['num-of-doors'].replace(np.nan, "four", inplace=True)
df.head(15)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_26676\2385977011.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['num-of-doors'].replace(np.nan, "four", inplace=True)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0             122.0         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

## Finally, let's drop all rows that do not have price data:


In [34]:
# simply drop whole row with NaN in "price" column
df.dropna(subset=["price"], axis=0, inplace=True)


# reset index, because we droped two rows
df.reset_index(drop=True, inplace=True)
df.head(15)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
9           2               192          bmw       gas        std   
10          0               192          bmw       gas        std   
11          0               188          bmw       gas        std   
12          0               188          bmw       gas        std   
13          1             122.0          bmw       gas        std   
14          0             122.0          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two        sedan          rwd           front       101.2  ...   
10         four        sedan          rwd           front       101.2  ...   
11          two        sedan          rwd           front       101.2  ...   
12         four        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       103.5  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           108         mpfi  3.50    2.80               8.8        101   
10          108         mpfi  3.50    2.80               8.8        101   
11          164         mpfi  3.31    3.19               9.0        121   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          209         mpfi  3.62    3.39               8.0        182   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5800       

<h2> Lets Check It</h2>

In [37]:
df.head(15)
for column in headers:
    print(column)
print(missing_data[column].value_counts())
print("")

symboling
normalized-losses
make
fuel-type
aspiration
num-of-doors
body-style
drive-wheels
engine-location
wheel-base
length
width
height
curb-weight
engine-type
num-of-cylinders
engine-size
fuel-system
bore
stroke
compression-ratio
horsepower
peak-rpm
city-mpg
highway-mpg
price
price
False    201
True       4
Name: count, dtype: int64



In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

Hurrey!!!
Now we have our dataset with no missing values.
<br>
Isn't It?

<br>
<br>

# REALLY IT WAS A GREAT JOB DUDE!!! 
<h4>LETS START LEARNING DATA FORMATING & DATA STANDARDIZATION.</h4>
<h4>There you will learn how to correct the data types of each columns in proper way. </h4>
<br>
<br>

# Bit ML
<h4>We make complex simple.</h4>
<h4>It's Easy If you learn like this.</h4>

# Thank you So much.